tower/src/make/make_service.rs

```diff
@@ -247,3 +247,80 @@ where
         self.make.make_service(target)
     }
 }
+
+/// A [`MakeService`] that produces services by cloning an inner service.
+#[derive(Debug, Clone, Copy)]
+pub struct Shared<S> {
+    service: S,
+}
+
+impl<S> Shared<S> {
+    /// Create a new [`Shared`] from a service.
+    pub fn new(service: S) -> Self {
+        Self { service }
+    }
+}
+
+impl<S> Service<()> for Shared<S>
```

Comment:
Couldn't this be generalized to any arbitrary target (that is just ignored)?

Reply:
It could but could that perhaps be confusing to users looking at the docs?

Reply:
IMO, I agree with @olix0r --- this should _probably_ be `Service<T>` for an arbitrary `T`, since that way, a `Shared` can be passed into functions expecting _any_ `MakeService`. The current impl can _only_ be passed intop functions expecting a `MakeService<()>`. 

If it's possible to do that easily, I think this would be better.

Reply:
Would we also not want to enforce that `S: impl Service`?

Reply:
@LucioFranco I don't think that's strictly necessary. You may have a Service that returns a discovery stream, for instance, or some non-service response. Also, that bound would require binding the request type via phantomdata, which seems better to avoid.

Reply:
Changing the impl to `Service<T>` leads to some inference errors in https://github.com/tower-rs/tower/blob/david/shared/tower/src/make/make_service.rs#L297:

```rust
error[E0282]: type annotations needed
   --> tower/src/make/make_service.rs:300:22
    |
300 |         poll_fn(|cx| MakeService::poll_ready(&mut shared, cx))
    |                      ^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `Target` declared on the trait `MakeService`

error[E0282]: type annotations needed
   --> tower/src/make/make_service.rs:314:26
    |
314 |         let mut shared = MakeService::into_service(shared);
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `Target` declared on the trait `MakeService`
````

Not sure if there is a way around those 🤔

Reply:
Other than calling `MakeService::<(), _>::poll_ready(&mut shared, cx)` of course, but that isn't very elegant.

Reply:
i'm not sure if the type inference issues are _that_ big a deal, because I imagine in practice (e.g. not in tests), these are mostly going to be passed into functions taking `T: MakeService<Whatever, Req>` or similar?